File: AI(ICE-3102)LAB/Tic-tac-Toe.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

char board[3][3];
int moves = 0;

// Function to initialize the board
void initializeBoard() {
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            board[i][j] = ' ';
}

// Function to print the board
void printBoard() {
    for (int i = 0; i < 3; i++) {
        printf("| %c | %c | %c |\n", board[i][0], board[i][1], board[i][2]);
        if (i < 2) printf("|---|---|---|\n");
    }
    printf("\n");
}

// Function to check for a winner
int checkWinner() {
    for (int i = 0; i < 3; i++) {
        if (board[i][0] == board[i][1] && board[i][1] == board[i][2] && board[i][0] != ' ')
            return 1;
        if (board[0][i] == board[1][i] && board[1][i] == board[2][i] && board[0][i] != ' ')
            return 1;
    }
    if (board[0][0] == board[1][1] && board[1][1] == board[2][2] && board[0][0] != ' ')
        return 1;
    if (board[0][2] == board[1][1] && board[1][1] == board[2][0] && board[0][2] != ' ')
        return 1;
    return 0;
}

// Function to check for a draw
int checkDraw() {
    return moves == 9;
}

// Function to make the human move
void humanMove() {
    int move;
    printf("Enter your move (1-9): ");
    scanf("%d", &move);
    int row = (move - 1) / 3;
    int col = (move - 1) % 3;

    if (board[row][col] == ' ') {
        board[row][col] = 'X';
        moves++;
    } else {
        printf("Invalid move! Try again.\n");
        humanMove();
    }
}

// Function to make the CPU move
void cpuMove() {
    int move;
    printf("Computer's turn!\n");
    do {
        move = rand() % 9 + 1;
        int row = (move - 1) / 3;
        int col = (move - 1) % 3;
        if (board[row][col] == ' ') {
            board[row][col] = 'O';
            moves++;
            break;
        }
    } while (1);
    printf("Computer chooses position %d.\n", move);
}

// Main game loop
int main() {
    srand(time(0));
    initializeBoard();
    while (1) {
        printBoard();
        humanMove();
        if (checkWinner()) {
            printBoard();
            printf("You win!\n");
            break;
        }
        if (checkDraw()) {
            printBoard();
            printf("It's a draw!\n");
            break;
        }

        cpuMove();
        if (checkWinner()) {
            printBoard();
            printf("Computer wins!\n");
            break;
        }
        if (checkDraw()) {
            printBoard();
            printf("It's a draw!\n");
            break;
        }
    }
    return 0;
}

File: Tic-Tac-Toe/Tic-tac-toe.py
import random

# Global variables
board = [[' ' for _ in range(3)] for _ in range(3)]
moves = 0

# Function to initialize the board
def initialize_board():
    global board, moves
    board = [[' ' for _ in range(3)] for _ in range(3)]
    moves = 0

# Function to print the board
def print_board():
    for i in range(3):
        print(f"| {board[i][0]} | {board[i][1]} | {board[i][2]} |")
        if i < 2:
            print("|---|---|---|")
    print("\n")

# Function to check for a winner
def check_winner():
    # Check rows and columns
    for i in range(3):
        if board[i][0] == board[i][1] == board[i][2] != ' ':
            return True
        if board[0][i] == board[1][i] == board[2][i] != ' ':
            return True
    # Check diagonals
    if board[0][0] == board[1][1] == board[2][2] != ' ':
        return True
    if board[0][2] == board[1][1] == board[2][0] != ' ':
        return True
    return False

# Function to check for a draw
def check_draw():
    return moves == 9

# Function to make the human move
def human_move():
    global moves
    while True:
        try:
            move = int(input("Enter your move (1-9): ")) - 1
            row, col = divmod(move, 3)
            if board[row][col] == ' ':
                board[row][col] = 'X'
                moves += 1
                break
            else:
                print("Invalid move! Try again.")
        except (ValueError, IndexError):
            print("Please enter a valid number between 1 and 9.")

# Function to make the CPU move
def cpu_move():
    global moves
    print("Computer's turn!")
    while True:
        move = random.randint(1, 9) - 1
        row, col = divmod(move, 3)
        if board[row][col] == ' ':
            board[row][col] = 'O'
            moves += 1
            print(f"Computer chooses position {move + 1}.")
            break

# Main game loop
def main():
    initialize_board()
    while True:
        print_board()
        human_move()
        if check_winner():
            print_board()
            print("You win!")
            break
        if check_draw():
            print_board()
            print("It's a draw!")
            break

        cpu_move()
        if check_winner():
            print_board()
            print("Computer wins!")
            break
        if check_draw():
            print_board()
            print("It's a draw!")
            break

if __name__ == "__main__":
    main()
